test/e2e/reconciler/namespace/controller_test.go

```diff
@@ -126,6 +136,100 @@ func TestNamespaceScheduler(t *testing.T) {
+					if err != nil {
+						klog.Errorf("failed to set status.lastSyncerHeartbeatTime: %v", err)
+						return
+					}
+				}, time.Second*10)
+
+				t.Log("Create a new unique sheriff CRD")
+				group := uuid.New().String() + ".io"
+				crd := newSheriffCRD(group)
+				gvr := schema.GroupVersionResource{Group: crd.Spec.Group, Resource: "sheriffs", Version: "v1"}
+				err = configcrds.CreateSingle(ctx, crdClusterClient.Cluster(server.clusterName).ApiextensionsV1().CustomResourceDefinitions(), crd)
+				require.NoError(t, err, "error bootstrapping CRD %s in cluster %s", crd.Name, server.clusterName)
+				require.Eventually(t, func() bool {
+					_, err := dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+					return err == nil
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to see CRD in cluster")
+
+				t.Log("Create a sheriff and wait for it to be scheduled")
+				_, err = dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Create(ctx, newSheriff(group, "woody"), metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create sheriff")
+				require.Eventually(t, func() bool {
+					obj, err := dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Get(ctx, "woody", metav1.GetOptions{})
+					if err != nil {
+						return false
+					}
+					require.NoError(t, err, "failed to get sheriff")
+					return obj.GetLabels()[clusterLabel] != ""
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to see sheriff scheduled")
+
+				t.Log("Delete the sheriff and the sheriff CRD")
+				err = dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Delete(ctx, "woody", metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete sheriff")
+				err = crdClusterClient.Cluster(server.clusterName).ApiextensionsV1().CustomResourceDefinitions().Delete(ctx, crd.Name, metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete CRD")
+
+				time.Sleep(30 * time.Second)
+
+				t.Log("Recreate the CRD, and then quickly a namespace, a configmap and a CR whose CRD was just recreated")
+				err = configcrds.CreateSingle(ctx, crdClusterClient.Cluster(server.clusterName).ApiextensionsV1().CustomResourceDefinitions(), crd)
+				require.NoError(t, err, "error bootstrapping CRD %s in cluster %s", crd.Name, server.clusterName)
+				_, err = kubeClusterClient.Cluster(server.clusterName).CoreV1().Namespaces().Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "namespace-test"}}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create namespace")
+				_, err = dynamicClusterClient.Cluster(server.clusterName).Resource(gvr).Namespace("default").Create(ctx, newSheriff(group, "lucky-luke"), metav1.CreateOptions{})
```

Comment:
Do we need to wait for either the API to show in discovery, or for the CRD conditions to be established? I'm worried this could flake (?)

Reply:
the bootstrapper does that, doesn't it?

Reply:
Oh, right, it makes sure it's established.